<font face="Helvetica" size=7 color='#2E0202'> **Analyzing** <font> <font face="Helvetica" size=7 color='#B40E0E'> **Yelp** <font> <font face="Helvetica" size=7 color='#2E0202'> **Customer Review Trends** <font>

<font face="Helvetica" size=4> **Group A4** <font>

*   
    Ari Kolahal
*   Wei-An (Vivian) Huang
*   Rinaldo Brendon Patel
*   Kendall Sims
*   Mingshu Hong
*   MingHua Tsai

<font face="Helvetica" size=4> **Table of Contents** <font>


1.<font face="Helvetica" size=3>  [Executive Summary](#executive-summary)<font>

2.<font face="Helvetica" size=3>   [Project Proposal](#project-proposal)<font>
  * <font face="Helvetica" size=3> [Introduction](#introduction)<font>
  * <font face="Helvetica" size=3> [Motivation](#motivation) <font>

3.<font face="Helvetica" size=3> [Data & its Source](#data&itssource)<font>
* <font face="Helvetica" size=3>  [Data Dictonary](#datadictionary) <font>


4.<font face="Helvetica" size=3> [Data Cleaning and Manipulation](#DataCleaningandManipulation) <font>

5.<font face="Helvetica" size=3> [Exploratory Data Analysis](#exploratorydataanalysis)<font>

6.<font face="Helvetica" size=3> [Conclusions](#conclusions)

7.<font face="Helvetica" size=3> [Challenges](#challenges)


8.<font face="Helvetica" size=3> [Reference](#reference)

9.<font face="Helvetica" size=3> [Generative-AI-Disclosure](#genai) <font>

<font face="Helvetica" size=6 color=#B40E0E> **Executive Summary <a name="executive-summary"></a>** <font>
    
----
    

    
This study leverages real Yelp data, SQL, and Google Cloud tools (Cloud Storage, BigQuery, Vertex AI) to examine customer satisfaction trends that can help restaurant owners improve their Yelp ratings. Focusing on factors like service quality, menu variety, ambiance, and pricing, the analysis provides practical insights into enhancing customer experiences and targeting specific audiences based on regional and demographic patterns.

The analysis highlights a steady rise in Yelp reviews from 2005 to 2019, peaking in 2018 before a significant decline in 2020 due to COVID-19. Elite users play a prominent role, contributing nearly 400 reviews on average compared to 34 by non-elite users, and garnering higher "useful" votes, fans, and compliments, indicating their influence within the community. Additionally, wheelchair-accessible locations receive the highest ratings and review counts, suggesting that inclusivity and operational quality enhance customer perceptions and positive feedback.

Key recommendations focus on quality over quantity, encouraging restaurant owners to build strong customer loyalty by prioritizing high-quality service at a few locations, with expansion only once standards are consistently high. Yelp can enhance engagement by prompting users to add details to positive reviews, creating sections for "Top Cool Reviews" and "Funny Reviews" to make browsing more enjoyable. Highlighting "useful" or "cool" reviews could guide businesses in improving their services, while public recognition and badges for elite users would further motivate quality contributions on the platform.
    

<font face="Helvetica" size=6 color=B40E0E> **Project Proposal <a name="project-proposal"></a>** <font>


<font face="Helvetica" size=5 color=2E0202> **Introduction <a name="introduction"></a>** <font>
    
---
    
In today's world of online delivery and food delivery apps, established restaurants have access to a wealth of data that can help them improve and grow based on customer feedback. However, this abundance of information also creates challenges for current businesses in identifying key areas for improvement and may seem daunting to future restaurant owners, especially in a highly competitive market like the United States. To succeed, businesses need insightful analysis to better market themselves and enhance their ratings, as many people rely on online reviews to decide where to dine. In this study, we analyze real data from a popular customer review website,Yelp, to understand sentiment and satisfaction scores across the platform, offering insights that can help restaurant owners make informed decisions to elevate their establishments.

<font face="Helvetica" size=5, color = 2E0202> **Motivation <a name="motivation"></a>** <font>

---
    
The motivation behind this study is to understand how Yelp reviews shape consumer choices, providing restaurants with actionable insights into customer preferences and areas for improvement. In today’s competitive market, especially within the restaurant industry, customer feedback is a crucial tool for businesses striving to refine their offerings and enhance customer satisfaction.

By analyzing patterns in review frequency and consistency, we aim to identify trends in customer loyalty and reviewing behaviors over time. This can help restaurants understand what drives repeat customers and how to foster lasting relationships. Furthermore, we seek to uncover key factors contributing to the popularity of highly-rated businesses—such as optimal business hours, strategic location, and desirable amenities—giving insights into the specific attributes that appeal most to customers. Additionally, examining changes in ratings over time allows us to assess how shifts in customer feedback impact a restaurant's reputation and performance. This analysis not only highlights areas where businesses are excelling but also reveals potential areas for improvement. 
    
Ultimately, our study provides restaurants with the data-driven insights needed to adapt to customer sentiment, boost their ratings, and secure a competitive edge in the ever-evolving dining landscape.

<font face="Helvetica" size=6 color=B40E0E> **Data & its Source<a name="data&itssource"></a>** <font>

----
    
    


The data for this project comes from 1 source:

- [**Yelp Dataset**:](https://www.yelp.com/dataset)<font>
    - The dataset has 4 main tables:
        1. **Business Table** (information about the location, star count, no. of reviews, business hours etc.) - 150346 Rows, and 14 Columns
        2. **Reviews Tables** (type of review business recieved) - 6990280 Rows, 9 Columns
        3. **User Table** ( info about user's yelp review behavior ) - 1987897 Rows, 22 Columns
        4. **Tip Table** ( type of tip posted by user) - 150346 Rows, 14 Columns
    
At the end of the the cleaning process, the following represents the metadata:

<font face="Helvetica" size=5, color = 2E0202> **Data Dictionary <a name="datadictionary"></a>** <font>

---

<font face="Helvetica" size=4, color = 2E0202> **Business Table:<a name="motivation"></a>** <font>


Column | Description | Data Type
------ | ----------- | ---------
**business_id** | Business Unique ID | ``string``
**name** | Business Name | ``string``
**address** | Business address | ``string``
**city** | Business city | ``string``
**state** | Business state | ``string``
**postal_code** | Business postal code| ``string``
**latitude** | Latitude of business location | ``float``
**longitude** |  Longitude of business location | `float`
**stars** | overall business star rating | ``float``
**review_count** | Number of reviews recieved | ``integer``
**is_open** | Business open or shutdown indicator | `integer`
**categories** | Type of business | ``string``
**hours** | Hours open accross the week | ``string``
**attributes** | Facilities offered at the establishment | `string`


<font face="Helvetica" size=4, color = 2E0202> **Reviews Table:<a name="motivation"></a>** <font>


Column | Description | Data Type
------ | ----------- | ---------
**review_id** | Review Unique ID| ``string``
**business_id** | Business Unique ID | ``string``
**user_id** | User Unique ID | ``string``
**stars** | Star rating provided for the review | ``float``
**date** | Timestamp of the review | ``timestamp``
**useful** | Categorized review as useful| ``integer``
**funny** | Categorized review as funny | ``integer``
**cool** |  Categorized review as cool | `integer`


<font face="Helvetica" size=4, color = 2E0202> **Tip Table:<a name="motivation"></a>** <font>


Column | Description | Data Type
------ | ----------- | ---------
**business_id** | Business Unique ID | ``string``
**user_id** | User Unique ID | ``string``
**date** | Timestamp of the tip | ``timestamp``
**compliment_count** | Indicator of compliment given or not | ``integer``


<font face="Helvetica" size=4, color = 2E0202> **User Table:<a name="motivation"></a>** <font>


Column | Description | Data Type
------ | ----------- | ---------
**user_id** | User Unique ID | ``string``
**name** | User Name | ``string``
**elite** | Year when declared Elite User| ``float``
**review_count** | Number of reviews given | ``integer``
**yelping_since** | Timestamp of user account created | ``timestamp``
**funny** | Total number of funny reviews given| ``integer``
**useful** | Total number of useful reviews given| ``integer``
**cool** | Total number of cool reviews given| ``integer``
**compliment_list** | Number of compliments given | ``integer``
**friends** | user_id declared as friends | ``string``
**average_stars** | average star rating the user gives | `float`
**fans** | Number of fans | ``integer``
**compliment_hot** | Count of compliment type given | ``integer``
**compliment_cute** | Count of compliment type given | ``integer``
**compliment_more** | Count of compliment type given | ``integer``
**compliment_profile** | Count of compliment type given | ``integer``
**compliment_note** | Count of compliment type given | ``integer``
**compliment_plain** | Count of compliment type given | ``integer``
**compliment_cool** | Count of compliment type given | ``integer``
**compliment_funny** | Count of compliment type given | ``integer``
**compliment_photos** | Count of compliment type given | ``integer``
**compliment_writer** | Count of compliment type given | ``integer``


<font face="Helvetica" size=6 color=B40E0E> **Data Cleaning and Manipulation<a name="DataCleaningandManipulation"></a>** <font>

----
    

## Cleaning Business Table

#### Overview of current table

In [2]:
%%bigquery 
SELECT * EXCEPT(int64_field_0)
FROM `ba775-project-a04.yelp_data.business`
LIMIT 10;

Query is running:   0%|          |

Downloading:   0%|          |

business_id                                         name  \
0  AWG2QYOoydgg2hMMk3KRoA                               The Home Depot   
1  F87os-6oNyUAEeGxQJKFkw                                McLeod Tavern   
2  --7PUidqRWpRSpXebiyxTg                   Humpty's Family Restaurant   
3  xPmbr4S0_XYweiNmga9dNA            Bird's Nest of Beijing Restaurant   
4  jgUJ2CkM3LaRHpt3JhKc7g                                      Walmart   
5  qdqm5QwfF-5ZqQuFlLFtSg                                    Game City   
6  WzWBfCHurRUGpQuXkMnfow                                  Copa Cabana   
7  -s8G4_7aF8IxlAq6p1w4rw                                Mucho Burrito   
8  dxNmbVCP-HGrDkPi4lqKDw     Cherry Hill Motors-Mercedes-Benz Leasing   
9  onPYxtkOmTIGYJbLCPX9Qg  State of New Jersey Motor Vehicle Comission   

                                             address           city state  \
0                               Terwillegar Drive NW       Edmonton    AB   
1                                 10009 101 Avenue N       Edmonton    AB   
2                                   9910 108A Avenue       Edmonton    AB   
3             10425 100 Ave NW, Comfort Inn & Suites       Edmonton    AB   
4                             156 Street & 87 Avenue  West Edmonton    AB   
5  82 Avenue & 83 Street, Bonnie Doon Shopping Ce...       Edmonton    AB   
6                             8230 Gateway Boulevard       Edmonton    AB   
7                                    8505 109 Street       Edmonton    AB   
8                                1800 Marlton Pike W    Cherry Hill    NJ   
9                        1 Executive Campus, Ste 110    Cherry Hill    NJ   

  postal_code   latitude   longitude  stars  review_count  is_open  \
0        None  53.438895 -113.606727    2.0             5        1   
1         T5H  53.542774 -113.483476    2.0            13        0   
2         T5H  53.554659 -113.493040    2.0            12        0   
3         T5K  53.538694 -113.500775    2.0             9        0   
4         T5R  53.520337 -113.593785    2.0             9        1   
5         T6C  53.544389 -113.490927    2.0             5        1   
6         T6E  53.518848 -113.495416    2.0             5        0   
7         T6G  53.521370 -113.511592    2.0             8        1   
8       08002  39.919082  -75.031452    2.0             7        1   
9       08002  39.928463  -75.052599    2.0            10        1   

                                          attributes  \
0  {'RestaurantsPriceRange2': '2', 'BusinessParki...   
1  {'RestaurantsTakeOut': 'False', 'HappyHour': '...   
2  {'GoodForKids': 'True', 'RestaurantsAttire': "...   
3  {'RestaurantsGoodForGroups': 'True', 'OutdoorS...   
4  {'RestaurantsPriceRange2': '2', 'BikeParking':...   
5  {'RestaurantsPriceRange2': '3', 'Caters': 'Fal...   
6  {'NoiseLevel': "u'loud'", 'RestaurantsDelivery...   
7  {'RestaurantsAttire': "'casual'", 'Restaurants...   
8  {'Caters': 'False', 'RestaurantsTableService':...   
9  {'Caters': 'False', 'RestaurantsTableService':...   

                                          categories  \
0  Home & Garden, Fashion, Shopping, Department S...   
1  Restaurants, Canadian (New), Lounges, Nightlif...   
2                    Breakfast & Brunch, Restaurants   
3                               Restaurants, Chinese   
4  Grocery, Department Stores, Shopping, Food, Fa...   
5  Videos & Video Game Rental, Books, Mags, Music...   
6               Latin American, Restaurants, Mexican   
7                               Restaurants, Mexican   
8               Automotive, Auto Repair, Car Dealers   
9  Pizza, Restaurants, Public Services & Governme...   

                                               hours  
0                                               None  
1  {'Monday': '11:0-0:0', 'Tuesday': '11:0-0:0', ...  
2                                               None  
3                                               None  
4  {'Monday': '8:0-22:0', 'Tuesday': '8:0-22:0', ...  
5               

#### Creating a new table filtering the businesses that are categorized as Restaurants or Food Establishments

In [3]:
%%bigquery
CREATE OR REPLACE TABLE
  `ba775-project-a04.yelp_data.restaurant_businesses` AS
      
SELECT *
FROM
  `ba775-project-a04.yelp_data.business`
WHERE
  categories LIKE '%Restaurant%'
    AND categories LIKE '%Food%'
    AND stars IS NOT NULL;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

#### Checking if any restaurants do not have a review

In [4]:
%%bigquery
SELECT B.business_id, R.business_id, R.review_id
FROM `ba775-project-a04.yelp_data.restaurant_businesses` AS B
LEFT JOIN `ba775-project-a04.yelp_data.review` AS R
ON B.business_id = R.business_id
WHERE R.review_id IS NUll;

Query is running:   0%|          |

Downloading: |          |

Empty DataFrame
Columns: [business_id, business_id_1, review_id]
Index: []

## Cleaning Tip Table

#### Overview of current table

In [5]:
%%bigquery
SELECT * FROM `ba775-project-a04.yelp_data.tip`

Query is running:   0%|          |

Downloading:   0%|          |

date  compliment_count             business_id  \
0      2009-07-30 15:13:31+00:00                 0  --0iUa4sNDFiZFrAdIWhZQ   
1      2014-07-08 01:00:55+00:00                 0  --0iUa4sNDFiZFrAdIWhZQ   
2      2012-09-07 20:29:44+00:00                 0  --0iUa4sNDFiZFrAdIWhZQ   
3      2011-07-18 22:30:31+00:00                 0  --0iUa4sNDFiZFrAdIWhZQ   
4      2010-09-13 21:43:09+00:00                 0  --0iUa4sNDFiZFrAdIWhZQ   
...                          ...               ...                     ...   
908910 2015-01-14 05:08:28+00:00                 4  NszFe9FGonnbHtGA2A_-Yw   
908911 2018-06-19 18:57:10+00:00                 4  kaQ-ReAmU-xO2XnQsMFwRg   
908912 2020-01-17 14:15:20+00:00                 5  pPpaOXOwcuO7z0sDghmOgw   
908913 2020-03-05 01:28:45+00:00                 6  dsfRniRgfbDjC8os848B6A   
908914 2016-06-11 15:37:07+00:00                 6  x8-sTKZG59RUhgGj_kcyVg   

                       user_id  
0       DPSY6qV8RmQZ_XlIjIw_SA  
1       LAwE-z6LRwUNg7-pR6DeDQ  
2       j_orXueAYJBft-VUNbs1IA  
3       mswloq-IjRr9yXLhBqBPAw  
4       mswloq-IjRr9yXLhBqBPAw  
...                        ...  
908910  Vg-ObmMxN7bi7fS__9P7vg  
908911  tsMF0FcFcHZ8i28WzWtQXw  
908912  tsMF0FcFcHZ8i28WzWtQXw  
908913  tsMF0FcFcHZ8i28WzWtQXw  
908914  661RwsBrt5ZbNhuipyhJcQ  

[908915 rows x 4 columns]

#### Checking if any null values exist in the data

In [6]:
%%bigquery
SELECT *
FROM `ba775-project-a04.yelp_data.tip`
WHERE user_id IS NULL
   OR business_id IS NULL
   OR compliment_count IS NULL
   OR date IS NULL;

Query is running:   0%|          |

Downloading: |          |

Empty DataFrame
Columns: [date, compliment_count, business_id, user_id]
Index: []

#### Checking if the date variable can be expanded upon

In [7]:
%%bigquery
SELECT
    date,
    compliment_count,
    business_id,
    user_id,
    EXTRACT(YEAR FROM date) AS year,
    EXTRACT(MONTH FROM date) AS month
FROM
    `ba775-project-a04.yelp_data.tip`
LIMIT 1000;

Query is running:   0%|          |

Downloading:   0%|          |

date  compliment_count             business_id  \
0   2021-12-03 18:35:53+00:00                 0  ---kPU91CF4Lq2-WlRu9Lw   
1   2021-01-12 23:10:44+00:00                 0  ---kPU91CF4Lq2-WlRu9Lw   
2   2021-03-30 03:53:21+00:00                 0  ---kPU91CF4Lq2-WlRu9Lw   
3   2021-09-23 16:58:31+00:00                 0  ---kPU91CF4Lq2-WlRu9Lw   
4   2016-12-10 19:34:56+00:00                 0  --30_8IhuyMHbSOcNWd6DQ   
..                        ...               ...                     ...   
995 2011-06-26 20:53:09+00:00                 0  -85kJMtb9wqNWDT8yLbitw   
996 2016-03-22 23:44:56+00:00                 0  -86Z04IBKxhEQ17rCOkn8g   
997 2014-06-13 18:03:38+00:00                 0  -86lWqykjlPF2MBrHRFcCw   
998 2012-12-27 21:33:13+00:00                 0  -8NGFqFrlOmWawr827tZNg   
999 2013-03-19 23:23:51+00:00                 0  -8NGFqFrlOmWawr827tZNg   

                    user_id  year  month  
0    zmgsdGzOp08BWJZ2yrOKww  2021     12  
1    VjvrRYcqu34DyBns0R11bg  2021      1  
2    VjvrRYcqu34DyBns0R11bg  2021      3  
3    St3aCytjxiMN-5OvFvcgSw  2021      9  
4    jd0gf26c8aERhlXnwTwueQ  2016     12  
..                      ...   ...    ...  
995  RXC4vbHOgwfYWmj0CxOAyw  2011      6  
996  A9pc0rdtWY5BY3LDHlVdGA  2016      3  
997  PDjTbAflKSqLZFHKgsh9fg  2014      6  
998  kYhCT0gkg1ZvZko4d3dDIQ  2012     12  
999  Znw0YbrA7hXFDQ0ricdS8g  2013      3  

[1000 rows x 6 columns]

#### Filtering out tips relevant to restaurant businesses

In [9]:
%%bigquery
CREATE TABLE `ba775-project-a04.yelp_data.restaurant_tip` AS
SELECT T.*
FROM `ba775-project-a04.yelp_data.restaurant_businesses` AS B
LEFT JOIN `ba775-project-a04.yelp_data.tip` AS T
USING(business_id);

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

## Cleaning Review Table

#### Overview of current table

In [10]:
%%bigquery
SELECT * FROM `ba775-project-a04.yelp_data.review`
LIMIT 10

Query is running:   0%|          |

Downloading:   0%|          |

cool  stars                      date  funny               review_id  \
0     0    2.0 2019-07-10 19:57:25+00:00      0  055lUC4jB4QAIViINcEJxg   
1     0    2.0 2017-01-10 18:37:09+00:00      0  Bg9aQH_cF8cj3BvU8aNFLw   
2     0    2.0 2015-10-11 00:24:00+00:00      0  rxlfJ9MvM4R-iAosqTQsrg   
3     0    2.0 2018-03-09 17:36:42+00:00      0  iRIPSMXJxMCIWrAh14cB3A   
4     0    2.0 2017-11-28 01:29:40+00:00      0  m5WzAQ3YlNgFgkMUJ8oIfA   
5     0    2.0 2019-02-15 21:55:24+00:00      0  VXU99P2GpDsfTFd4MJ8YLw   
6     0    2.0 2019-07-20 15:15:49+00:00      0  ePba2sXrWEys5YPoCrUetA   
7     0    2.0 2016-01-05 16:48:55+00:00      0  wHVlwrB-zHIp-bPN6ZiDUA   
8     0    2.0 2020-02-16 15:08:35+00:00      0  zqhBFUnipnVGHWBhp-SaOg   
9     0    2.0 2016-09-06 20:47:27+00:00      0  Wj44GcP98vM2O7VrpYJvaQ   

   useful             business_id                 user_id  
0       0  DFZnh7ApQMk0nYvwTlRUxw  hNDZGIu9Aovra5VGVk1heQ  
1       0  KyjOahjnvnz28iPsXddJow  G2J7KkH1Y2qAdUM3wbJlQA  
2       0  tJL-K8vopI-mN3eFW9YnJQ  2de2hu_AdnR4wcifTlkkqA  
3       0  g-IXI3v3fOudMsyLPKb-eQ  miNT-FJ5WEFV-uk4pLOQnQ  
4       0  yK8LWJ4VRTYEoGxVYzcjQg  LbisQaSrSdJCnW6Q-XzeAA  
5       0  aAdGIK9ewmiWtGzIcrQgNA  0A46uacaUaFc9iTPe0L7CQ  
6       0  xJIeT5--AiPdvPb5BRflyQ  Eb7EcTHdbeWucQ6XA80aOw  
7       0  5Ubo-2HLRnCCo8Pnq-v-cQ  aSkP_BL_DRexxTOUka_doQ  
8       0  ASKMKkejMCoFA4gIfdhbNw  yeeqo5hRlSYEWH0x4KRHzA  
9       0  mEam9yl2KQs0ebKgj3mPZw  f0LcLqwZr29JOvlDw98JqA

#### Checking for any null values


In [11]:
%%bigquery
SELECT *
FROM `ba775-project-a04.yelp_data.review`
WHERE cool IS NULL
   OR stars IS NULL
   OR date IS NULL
   OR funny IS NULL
   OR review_id IS NULL
   OR useful IS NULL
   OR business_id IS NULL
   OR user_id IS NULL

Query is running:   0%|          |

Downloading: |          |

Empty DataFrame
Columns: [cool, stars, date, funny, review_id, useful, business_id, user_id]
Index: []

#### Checking if the date variable can be expanded upon

In [12]:
%%bigquery
SELECT
    cool,
    stars,
    date,
    funny,
    review_id,
    useful,
    business_id,
    user_id,
    EXTRACT(YEAR FROM date) AS year,
    EXTRACT(MONTH FROM date) AS month
FROM `ba775-project-a04.yelp_data.review`
LIMIT 10;

Query is running:   0%|          |

Downloading:   0%|          |

cool  stars                      date  funny               review_id  \
0     0    2.0 2013-09-12 23:25:41+00:00      0  Z0lgOjVAnZbYAdx7374KUA   
1     0    2.0 2011-11-18 14:00:09+00:00      0  bLWbL8iy8g_j_7USbKgeWw   
2     0    2.0 2015-03-27 00:07:49+00:00      0  -pepYYLPEZzEJekHEF1c-Q   
3     0    2.0 2018-04-21 12:49:28+00:00      0  xEfT0xY_cCIW-lxEsyRN8Q   
4     0    2.0 2021-06-18 02:40:55+00:00      0  T3UpHs1Fr8Ozu5OacUBBPw   
5     0    2.0 2017-04-21 15:58:53+00:00      0  UAjDSoYhOr4V1OODvAq08w   
6     0    2.0 2013-08-31 08:48:11+00:00      0  6HAp8Gp89FZeyIQmfJlwTw   
7     0    2.0 2012-02-21 00:37:14+00:00      0  SgP8aeBgepGbthxBF8kzZA   
8     0    2.0 2015-04-25 02:28:41+00:00      0  oriaeJhVX4REbkZ33gGwIQ   
9     0    2.0 2018-10-25 02:28:32+00:00      0  2uXnuKjiN90AcOEaG6Ji2Q   

   useful             business_id                 user_id  year  month  
0       0  e8IsHdOqg-MyytZ3-10jbA  NfoAmvEZ3NVTAtOEshCvwQ  2013      9  
1       0  0l_njS4wGMadAk1b4nZRUA  vbEf26yXYP0WBsveEXWIlw  2011     11  
2       0  CgiwCrXs3Ggu9eKFPIgjtA  eacSi8XUcBLXlDIuLMDsow  2015      3  
3       0  ktM6aYuzqWd-kA8SIWswiQ  IpLRJY4CP3fXtlEd8Y4GFQ  2018      4  
4       0  _ouzokto3qEMm1Xa3W1drw  gfdDF6W076x1B83OWyvX_w  2021      6  
5       0  f84O7bzOeLJbNphTn6cNHw  tdoFAdyE3wMAZ1K2tgtTOQ  2017      4  
6       0  TCWnTa69sJ3vGQsdEe3Esw  X5_3x28U9dxzpbh-KGkzdw  2013      8  
7       0  JP4jIbkthYZVKJCvMiZUDw  G-uv9J_2-0OyeIlNfHuYjw  2012      2  
8       0  3oy9A7cOIaQsuxrtmjyxIw  vsikhxSLtMk5APIf9U7C0w  2015      4  
9       0  JgTSG4rIDf20Kgp6U_asng  OKEiKXq8Q0LJFwc4bavH7w  2018     10

#### Filtering out reviews that are relevant for restaurant businesses

In [14]:
%%bigquery
CREATE TABLE `ba775-project-a04.yelp_data.restaurant_review` AS
SELECT R.*
FROM `ba775-project-a04.yelp_data.restaurant_businesses` AS B
LEFT JOIN `ba775-project-a04.yelp_data.review` AS R
USING(business_id);

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

## Cleaning User Table

#### Overview of current table

In [15]:
%%bigquery
SELECT *
FROM `ba775-project-a04.yelp_data.user`
LIMIT 10;

Query is running:   0%|          |

Downloading:   0%|          |

compliment_writer  compliment_photos  compliment_funny  compliment_cool  \
0                  0                  0                 0                0   
1                  0                  0                 0                0   
2                  0                  0                 0                0   
3                  0                  0                 0                0   
4                  0                  0                 0                0   
5                  0                  0                 0                0   
6                  0                  0                 0                0   
7                  0                  0                 0                0   
8                  0                  0                 0                0   
9                  0                  0                 0                0   

   compliment_plain  compliment_note  compliment_profile  compliment_more  \
0                 0                0                   0                0   
1                 0                0                   0                0   
2                 0                0                   0                0   
3                 0                0                   0                0   
4                 0                0                   0                0   
5                 0                0                   0                0   
6                 0                0                   0                0   
7                 0                0                   0                0   
8                 0                0                   0                0   
9                 0                1                   0                0   

   compliment_cute  compliment_hot  ...  average_stars  \
0                0               0  ...            5.0   
1                0               0  ...            5.0   
2                0               0  ...            5.0   
3                0               0  ...            4.0   
4                0               0  ...            1.0   
5                0               0  ...            3.0   
6                0               0  ...            1.0   
7                0               0  ...            5.0   
8                0               0  ...            5.0   
9                0               0  ...            5.0   

                                             friends  cool compliment_list  \
0  TL9LfHzDsBQq0y8ZDKKvzg, U0wyBwg45hDAzcUngmWGrA...     0               0   
1  CC5UNVSVSeOoasHUV5peTg, ofJi6IjAZYu3UXwfoX-9vA...     0               0   
2  m6sCpFCuEg9CdZ0LIe0WBg, YMQQ9Nf0pGdr2mQAjF6Mfw...     0               0   
3  g8O0f3INKP9aWhUoYK-IWg, BU71n38CovBLErKAuDF2pA...     0               0   
4  nbqd5Bbmd-orIHY7IKelvg, NhyfgaABV54xtSs23IDJdw...     0               0   
5  QAClo0cwSQuQnjtQDWmjtQ, N_RSb3sAX5SboWFW_l16rA...     0               0   
6  Tsw_GGovH0KMJ9qtPdJdYA, RAuFa1EhjHIrECuXElvrnQ...     0               0   
7  fa-w9QJs3kPMpnzyzfJXoQ, TPHTZHFLlcGF6u09NgqR3g...     0               0   
8  5wFTRAX-I2p1fzF1pZ4lng, b_tA7ch6jFPfKB-PIfeWww...     0               0   
9  U2I_5L2AMCi9_pkCuZ2IwA, zJP7YRH0f9NnJi0q-5EjJQ...     0               0   

   useful  funny             yelping_since            name review_count elite  
0       0      0 2012-10-17 12:10:17+00:00  GabeandRebecca            1   NaN  
1       0      0 2010-07-31 22:15:33+00:00          marcus            1   NaN  
2       0      0 2016-02-27 05:54:12+00:00          Daniel            1   NaN  
3       0      0 2010-10-21 19:09:12+00:00           Jason            1   NaN  
4       0      0 2012-02-21 21:31:26+00:00            Erin            1   NaN  
5       0      0 2013-02-10 01:13:55+00:00             R P            1   NaN  
6       0      0 2014-07-01 00:31:56+00:00            Nina            1   NaN  
7       0      0 2013-02-08 00:06:46+00:00            Nate            1   NaN  
8       0      0 2017-06-11 21:33:35+00:00        Savannah            1   NaN 

#### Checking to see how many rows have NULL elite years

In [16]:
%%bigquery
SELECT 
    user_id,
    ARRAY_TO_STRING(REGEXP_EXTRACT_ALL(CAST(elite AS STRING), r'20\d{2}'), ',') AS elite_years
FROM  
    `ba775-project-a04.yelp_data.user`
WHERE 
    elite IS NULL;

Query is running:   0%|          |

Downloading:   0%|          |

user_id elite_years
0        c27mcOhJ3U95601Eg7TzEg        None
1        GIY18XDDstyTcDgqNmyUjA        None
2        xQb6OJkROP0FsSsMFeaMQw        None
3        wK4dZnyS3_yl8ZJYT-gEwQ        None
4        aVzOv3aTr3SD9U5mKPFukA        None
...                         ...         ...
1896694  2Fw8kben7xD6ezDJxDjRSA        None
1896695  wMRABJye-vljUZxI2y7YTA        None
1896696  QslrEbjulbQl-gEw6oGxVw        None
1896697  CUExtAp3Blvq9uIf7yuAvg        None
1896698  PMXZP8zunHtM36C7p3Y_kw        None

[1896699 rows x 2 columns]

#### Filter out users that have left reviews for restaurant businesses

In [17]:
%%bigquery
CREATE TABLE `ba775-project-a04.yelp_data.restaurant_user` AS
SELECT U.*
FROM `ba775-project-a04.yelp_data.restaurant_businesses` AS B
LEFT JOIN `ba775-project-a04.yelp_data.review` AS R
USING(business_id)
LEFT JOIN `ba775-project-a04.yelp_data.user` AS U
USING(user_id);

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

<font face="Helvetica" size=6 color=B40E0E> **Exploratory Data Analysis<a name="exploratorydataanalysis"></a>** <font>

----
    

<font face="Helvetica" size=5 color=2E0202> **Q1. What are the general trends in Yelp restaurant reviews?** <font>

In [18]:
%%bigquery

SELECT DISTINCT EXTRACT(year from date) AS year, COUNT(review_id) AS review_count
FROM `ba775-project-a04.yelp_data.restaurant_review` 
GROUP by year
Order by year;

Query is running:   0%|          |

Downloading:   0%|          |

year  review_count
0   2005           132
1   2006           737
2   2007          2967
3   2008          9265
4   2009         15535
5   2010         29670
6   2011         51422
7   2012         64504
8   2013         89144
9   2014        124323
10  2015        169487
11  2016        188066
12  2017        203556
13  2018        224827
14  2019        222493
15  2020        135582
16  2021        152581
17  2022          7472

From 2005~2019, the number of annual reviews for restaurant reviews has shown a clear upward trend, indicating that more and more users are engaging with the platforms over time.The number of reviews started low in 2005, with only 132 reviews, and then grew significantly year over year, peaking in 2018 with 224,728 reviews. This trend suggests that restaurant review platforms are growing in popularity and that consumers may increasingly rely on user-generated reviews to make dining decisions.

However, starting in 2020, the number of reviews declines significantly, falling to 135,512 in 2020 and further to 7,466 by 2022. This steep decline is likely due to the impact of the COVID-19 pandemic, which led to lockdowns, restaurant closures, and less eating out, which in turn reduced reviews. Although there is a slight rebound in 2021, the number is still below the peak, indicating that user participation has not returned to pre-pandemic levels.


<font face="Helvetica" size=5 color=2E0202> **Q2. When did users start actively reviewing on Yelp?** <font>

In [19]:
%%bigquery
SELECT DISTINCT EXTRACT(year FROM yelping_since) AS year,
       COUNT(yelping_since) AS user_start_count
FROM `ba775-project-a04.yelp_data.restaurant_user` 
WHERE yelping_since IS NOT NULL
GROUP BY year
ORDER BY year

Query is running:   0%|          |

Downloading:   0%|          |

year  user_start_count
0   2004                41
1   2005              1745
2   2006              7860
3   2007             22858
4   2008             49507
5   2009             88169
6   2010            137195
7   2011            195431
8   2012            195419
9   2013            193200
10  2014            197129
11  2015            189839
12  2016            148849
13  2017             97389
14  2018             75920
15  2019             53883
16  2020             21653
17  2021             15077
18  2022               598

The data show the growth in registered users of the Platform from 2004 to 2022. Initially, the growth of the platform was modest, with only 41 new users in 2004. However, from 2005 onwards, the number of user registrations grew steadily and rapidly, peaking between 2011 and 2015, with the highest number of user registrations recorded in 2014 at approximately 197055. This surge likely reflects the growing popularity of the platform and increased interest in online reviews when making dining decisions.

Beginning in 2016, the number of new user registrations gradually declined, with a significant drop beginning in 2019. the drop was particularly pronounced in 2020, likely due to the COVID-19 pandemic limiting dining out activities and reducing consumer interest in restaurant reviews. By 2022, new user registrations fall to just 598, suggesting that engagement on the platform may still be recovering from the impact of the pandemic or may be turning to other mediums or platforms for restaurant-related information.


<font face="Helvetica" size=5 color=2E0202> **Q3. How active are elite users compared to non-elite users in terms of fans, compliments, reviews, and contributions?** <font>

In [21]:
%%bigquery
SELECT
   CASE WHEN elite IS NOT NULL THEN 'Elite' ELSE 'Non-Elite' END AS user_type,
   ROUND(AVG(review_count),2 ) AS avg_reviews,
   ROUND(AVG(useful), 2 ) AS avg_useful,
   ROUND(AVG(fans), 2) AS avg_fans,
   ROUND (AVG(
       compliment_hot + compliment_more + compliment_profile + compliment_cute +
       compliment_list + compliment_note + compliment_plain + compliment_cool +
       compliment_funny + compliment_writer + compliment_photos
   ), 2 ) AS avg_compliments,
   ROUND( AVG (average_stars), 2 ) AS avg_RatingOfAllReviews,
   ROUND (AVG(
       compliment_hot + compliment_more + compliment_profile + compliment_cute +
       compliment_list + compliment_note + compliment_plain + compliment_cool +
       compliment_funny + compliment_writer + compliment_photos + cool + useful + funny
   ), 2 )AS avg_interactions,
   COUNT(user_id) AS total_count
FROM `ba775-project-a04.yelp_data.restaurant_user`
GROUP BY user_type
ORDER BY user_type;


Query is running:   0%|          |

Downloading:   0%|          |

user_type  avg_reviews  avg_useful  avg_fans  avg_compliments  \
0      Elite       375.81     1477.12     47.12           594.79   
1  Non-Elite        36.64       39.50      1.11             6.31   

   avg_RatingOfAllReviews  avg_interactions  total_count  
0                    3.98           3712.20       460952  
1                    3.72             72.65      1230810

Elite users are members of the community recognized for their high-quality contributions. Yelp awards this designation to users who actively participate by writing helpful reviews, providing useful tips, and engaging positively with the Yelp community.
Elite users selected are those users who have been elite at least once during their time while being active reviewers. Following are the observations for elite users in comparison to the non-elite users from the query result.

- Elite users write significantly more reviews, with an average of 396.79 reviews per elite user than non-elite users who write around 34.37 reviews. 
- Elite user's reviews are valued in the community which is evident with the average useful votes of 1611.
- Elite users are popular on Yelp with 50 average fans, suggesting that elite users are frequently followed and their opinions are taken into account with high significance.
- Elite users show active participation and positive engagement as they are shown to receive significantly higher compliments on their reviews
- Both Elite and Non-Elite users have a similar average rating for their reviews, with elite users at 3.96 and non-elite users at 3.68, showing that rating behavior is somewhat consistent across user types.

The observations for elite users are seen as higher than non-elite is because elite users' reviews are highlighted, they are well known for writing quality reviews, and thus the extra amount of interactions in comparison to non-elite users


<font face="Helvetica" size=5 color=2E0202> **Q4. Which businesses have the most loyal customers, as seen through repeat reviews?** <font>

In [36]:
%%bigquery

SELECT B.name AS business_name, avg(R.stars) AS avg_stars, COUNT(R.user_id) AS loyal_user_occurance
FROM `ba775-project-a04.yelp_data.restaurant_businesses` AS B
INNER JOIN `ba775-project-a04.yelp_data.restaurant_review` AS R
USING(business_id)
GROUP BY B.name
HAVING COUNT(R.user_id)>1
ORDER BY loyal_user_occurance DESC
LIMIT 20;


Query is running:   0%|          |

Downloading:   0%|          |

business_name  avg_stars  loyal_user_occurance
0                           McDonald's   1.783249                 18196
1               Chipotle Mexican Grill   2.479054                  9763
2                            Taco Bell   2.071013                  8604
3                          Chick-fil-A   3.409406                  8378
4                    Acme Oyster House   4.124984                  7673
5                         Panera Bread   2.702915                  6483
6                       Steak ’n Shake   2.071418                  6455
7                              Wendy's   1.943255                  6397
8   Hattie B’s Hot Chicken - Nashville   4.446266                  6160
9              Reading Terminal Market   4.605400                  5778
10                         Burger King   1.946030                  5466
11                 Mother's Restaurant   3.438713                  5254
12           Popeyes Louisiana Kitchen   2.056346                  5200
13                           Starbucks   3.143903                  5191
14                        Jimmy John's   2.923279                  4562
15                           Five Guys   3.559400                  4537
16                      Sabrina's Café   4.133197                  4407
17                                Wawa   3.313622                  4346
18    Cracker Barrel Old Country Store   3.150242                  4333
19                Pat's King of Steaks   3.240158                  4293

The analysis reveals that McDonald's has the most loyal customers, evidenced by a significant number of repeat reviews from the same users, totaling 18,196 reviews. Following McDonald's, other chains show approximately 40% fewer reviews. Notably, the restaurants with the most loyal customers are largely large, nationwide chains in the USA. Interestingly, these chains tend to have relatively low average star ratings (below 3 stars). One possible explanation is that customers may be leaving frequent negative reviews as a way to express dissatisfaction and caution future diners about their experiences.

In contrast, restaurants with higher quality ratings (above 4 stars) are generally smaller, non-chain businesses operating in only a few locations. Their higher ratings may be attributed to their ability to maintain a higher quality of service and attention to detail, focusing on a localized area. Additionally, these smaller establishments may be more responsive to customer feedback, actively addressing issues to improve the dining experience.



<font face="Helvetica" size=5 color=2E0202> **Q5. What kind of restaurants with high review
counts have received compliments or tips?** <font>

In [23]:
%%bigquery
SELECT 
  B.business_id,
  B.name,
  B.review_count,
  B.stars,
  COUNT(T.compliment_count) AS received_compliments_count
FROM 
  `ba775-project-a04.yelp_data.restaurant_businesses` AS B
LEFT JOIN 
  `ba775-project-a04.yelp_data.restaurant_tip` AS T
ON 
  B.business_id = T.business_id
WHERE 
  B.review_count > 0
GROUP BY 
  B.business_id, B.name, B.review_count, B.stars
HAVING 
  received_compliments_count > 0
ORDER BY 
  received_compliments_count DESC;


Query is running:   0%|          |

Downloading:   0%|          |

business_id                                name  \
0      FEXhWNCMkv22qG04E83Qjg                       Café Du Monde   
1      _ab50qdWOk0DdB6XOrBitw                   Acme Oyster House   
2      ytynqOUb3hjKeJfRj5Tshw             Reading Terminal Market   
3      iSRTaT9WngzB8JJ2YKJUig                 Mother's Restaurant   
4      GXFMD0Z4jEVZBCsbPf4CTQ  Hattie B’s Hot Chicken - Nashville   
...                       ...                                 ...   
18744  siZvbtz6_oQkNiDVTjeYYA           24 Hours Pizza and Donair   
18745  WklNC4oBYyPqMb2ILfcV3w                          McDonald's   
18746  Ai8MnqWUCTT03knlsrDSOQ                          McDonald's   
18747  2GUTsH6BHqasxtBEcXwNzQ                          McDonald's   
18748  dD8HFOlOowiEcBv5F38pew                          McDonald's   

       review_count  stars  received_compliments_count  
0              1880    4.0                        2571  
1              7568    4.0                         932  
2              5721    4.5                         827  
3              5185    3.5                         606  
4              6093    4.5                         579  
...             ...    ...                         ...  
18744             5    1.5                           1  
18745             9    1.5                           1  
18746            11    1.5                           1  
18747            14    1.5                           1  
18748             9    1.5                           1  

[18749 rows x 5 columns]

We merged the yelp_business and yelp_tips tables from the dataset to compare and analyze the relationship between review_count and compliment_count. Our aim was to further explore which regions or types of restaurants tend to receive more reviews and to assess the proportion of positive feedback. From the data, we can observe that review_count and compliment_count are not necessarily directly correlated—some restaurants have a high review_count due to negative reviews. However, the majority of restaurants with a higher number of reviews tend to have a high compliment_count and stars rating.

Further analysis of the locations and types of these restaurants shows that LA, PA, and FL have the highest number of restaurants receiving compliments and reviews, with LA standing out at 24 restaurants. Next, we will compare these results with the total number of restaurants in each area to help restaurant owners develop more effective marketing strategies.

In terms of category, all types of restaurants have the opportunity to receive reviews, without being limited to any single type. The distribution is relatively even, suggesting that restaurant type is not a key factor influencing the number or positivity of reviews.


<font face="Helvetica" size=5 color=2E0202> **Q6. What are the seasonal and day-of-week trends in restaurant reviews on Yelp?** <font>

In [45]:
%%bigquery
# Seasonality - num of reviews per month
SELECT COUNT(review_id) AS num_reviews, EXTRACT(MONTH FROM date) AS month
FROM `ba775-project-a04.yelp_data.restaurant_review` 
GROUP BY month
ORDER BY num_reviews DESC;


Query is running:   0%|          |

Downloading:   0%|          |

num_reviews  month
0        158902      7
1        154660      8
2        145867      1
3        145854      6
4        145178      3
5        141792      5
6        138206     10
7        137048      9
8        133784      4
9        132008      2
10       129253     11
11       129211     12

Our dataset reveals that the highest number of Yelp reviews are consistently posted in July, August, and January. This trend may reflect heightened dining activity during these months; however, it does not suggest a clear seasonal pattern, as review volumes appear unaffected by general seasonal shifts (e.g., winter, spring, summer, fall). One possible reason for January’s popularity is the winter break for high school and college students, combined with the New Year holiday season when many people are on vacation and traveling. Increased travel often leads to more dining out, which could drive up Yelp review activity. Similarly, the popularity of reviews in July and August aligns with summer vacation, when people have more free time to dine out.

These findings offer valuable insights into how holidays and school schedules may impact Yelp review activity, and they suggest a clear direction for further analysis. Recognizing January, July, and August as peak months for Yelp reviews could help restaurants anticipate and prepare for high-demand periods, ultimately improving their service and customer satisfaction.

In [30]:
%%bigquery
# Seasonality - num of reviews per day of the week
SELECT
   COUNT(review_id) AS num_reviews,
   FORMAT_DATE('%A', date) AS day_of_week
FROM
   `ba775-project-a04.yelp_data.restaurant_review` 
GROUP BY
   day_of_week
ORDER BY
   num_reviews DESC;

Query is running:   0%|          |

Downloading:   0%|          |

num_reviews day_of_week
0       295833      Sunday
1       267939    Saturday
2       246547      Monday
3       223752      Friday
4       221728   Wednesday
5       220191     Tuesday
6       215773    Thursday

Our dataset reveals that Sunday is the most popular day for posting Yelp reviews, aligning with the trend of increased dining activity over the weekend. Interestingly, Saturday is the second most popular day for reviews, further confirming heightened weekend dining activity. Monday ranks as the third most popular day, possibly reflecting a pattern where people share their weekend dining experiences or opt to dine out as the week begins.

For restaurants, this insight underscores the importance of being well-staffed and efficient on weekends to manage increased customer volume effectively. Providing faster, reliable service on these high-traffic days can encourage positive reviews, boost customer satisfaction, and enhance online reputation.

<font face="Helvetica" size=5 color=2E0202> **Q7. Is there a relationship between the number of reviews a business has and its star rating?** <font>

In [32]:
%%bigquery
SELECT
   ROUND(AVG(CASE WHEN total_reviews > 2000 THEN average_rating END) , 2 ) AS More_than_2000_reviews,
   ROUND(AVG(CASE WHEN total_reviews >= 1000 THEN average_rating END), 2 ) AS More_than_1000_reviews,
   ROUND(AVG(CASE WHEN total_reviews >= 500 THEN average_rating END), 2 ) AS More_than_500_reviews,
   ROUND(AVG(CASE WHEN total_reviews <= 500 THEN average_rating END), 2 ) AS Less_than_500_reviews,
   ROUND(AVG(CASE WHEN total_reviews <= 200 THEN average_rating END), 2 ) AS Less_than_200_reviews,
   ROUND(AVG(CASE WHEN total_reviews <= 100 THEN average_rating END), 2 ) AS Less_than_100_reviews
FROM (
   SELECT business_id,
          COUNT(review_id) AS total_reviews,
          AVG(stars) AS average_rating
   FROM `ba775-project-a04.yelp_data.restaurant_review`
   GROUP BY business_id
);


Query is running:   0%|          |

Downloading:   0%|          |

More_than_2000_reviews  More_than_1000_reviews  More_than_500_reviews  \
0                    4.16                    4.15                   4.03   

   Less_than_500_reviews  Less_than_200_reviews  Less_than_100_reviews  
0                   3.43                   3.39                   3.33

The query describes the relationship between the number of reviews and the star rating. First, it takes count of the total reviews received for each business and the average star rating in those reviews. Post which the average star rating is calculated for all the restaurants that have more than 2000, 1000, and 500 reviews. To view the relationship between low reviews and star ratings, the average rating of restaurants that have less than 500, 200, and 100 reviews are taken into account. The output displays all these columns. 

The query results show that restaurants with more than 2000 reviews have an average rating of 4.15, more than 1000 have 4.14, and more than 500 have 4 as their average rating. Restaurants with less than 500 reviews have 3.42, less than 200 review restaurants have 3.38 and less than 100 review restaurants have 3.33 as their average review rating. 
This suggests that restaurants with large numbers of reviews seem to be receiving higher ratings. Users usually leave positive & critical reviews for the restaurants when they are satisfied with the restaurants. Good restaurants see more number of reviews. Restaurants with fewer reviews seem to have lower ratings suggesting mediocre restaurants don’t receive more reviews and if they do their reviews are lower, negative & critical reviews from the users. 


<font face="Helvetica" size=5 color=2E0202> **Q8. How do star ratings vary with review types (funny, cool, useful) for top-rated vs. low-rated businesses?** <font>

In [41]:
%%bigquery
SELECT
    stars,
    AVG(funny) AS avg_funny,
    AVG(cool) AS avg_cool,
    AVG(useful) AS avg_useful,
    COUNT(*) AS review_count
FROM
    `ba775-project-a04.yelp_data.restaurant_review`
GROUP BY stars
ORDER BY stars;

Query is running:   0%|          |

Downloading:   0%|          |

stars  avg_funny  avg_cool  avg_useful  review_count
0    1.0   0.380913  0.142265    1.023073        208470
1    2.0   0.394908  0.251981    1.145776        134796
2    3.0   0.365974  0.449201    1.051440        183379
3    4.0   0.355023  0.726291    1.133947        392918
4    5.0   0.252618  0.573082    0.900487        772200

In [44]:
%%bigquery
WITH ReviewCategories AS (
    SELECT
        review_id,
        stars,
        funny,
        cool,
        useful,
        CASE
            WHEN stars >= 4 THEN 'Top Rated'
            WHEN stars <= 2 THEN 'Low Rated'
            ELSE 'Neutral'
        END AS rating_category
    FROM
        `ba775-project-a04.yelp_data.restaurant_review`
)
SELECT
    rating_category,
    AVG(funny) AS avg_funny,
    AVG(cool) AS avg_cool,
    AVG(useful) AS avg_useful,
    COUNT(*) AS review_count
FROM
    ReviewCategories
WHERE
    rating_category IN ('Top Rated', 'Low Rated', 'Neutral')
GROUP BY
    rating_category
ORDER BY
    rating_category;


Query is running:   0%|          |

Downloading:   0%|          |

rating_category  avg_funny  avg_cool  avg_useful  review_count
0       Low Rated   0.386409  0.185349    1.071257        343266
1         Neutral   0.365974  0.449201    1.051440        183379
2       Top Rated   0.287153  0.624750    0.979218       1165118

The data shows that 1- and 2-star reviews are the funniest, while 4-star reviews are the "coolest," and 1-star reviews are the most "useful." When categorized into groups, low-rated reviews (1-2 stars) are the funniest and most useful, as they often include humorous or detailed critiques. Neutral reviews (3 stars) stand out as the "coolest," offering a balanced, relatable appeal. In contrast, top-rated reviews (4-5 stars), though positive, are generally seen as less helpful. This indicates that lower-rated reviews provide more specific feedback, making them both entertaining and informative, while neutral and top-rated reviews offer relatability and positivity, respectively.

Users might often find low-rated reviews funny because they include sarcasm or exaggerated complaints, making them entertaining to read. These low-rated reviews are also the most "useful," as they provide detailed, critical feedback that helps others make informed choices. Neutral (3-star) reviews are perceived as "cool" because they offer a balanced, relatable perspective. In contrast, highly positive reviews, while uplifting, tend to be less detailed and are seen as less helpful.


<font face="Helvetica" size=5 color=2E0202> **Q9. Which areas have the most closed vs. open businesses, and how do star ratings compare?** <font>

In [34]:
%%bigquery
SELECT 
  city,
  state,
  COUNT(business_id) AS open_business_count,
  AVG(stars) AS average_stars_open
FROM 
  `ba775-project-a04.yelp_data.restaurant_businesses`
WHERE 
  is_open = 1
GROUP BY 
  city, state
ORDER BY 
  open_business_count DESC
LIMIT 20;

Query is running:   0%|          |

Downloading:   0%|          |

city state  open_business_count  average_stars_open
0       Philadelphia    PA                 1465            3.566894
1              Tampa    FL                  961            3.534339
2       Indianapolis    IN                  906            3.229029
3             Tucson    AZ                  830            3.374096
4          Nashville    TN                  752            3.449468
5           Edmonton    AB                  623            3.387640
6        New Orleans    LA                  611            3.710311
7               Reno    NV                  382            3.484293
8        Saint Louis    MO                  370            3.441892
9              Boise    ID                  307            3.534202
10     Santa Barbara    CA                  194            3.940722
11        Clearwater    FL                  189            3.394180
12         St. Louis    MO                  188            3.236702
13          Metairie    LA                  165            3.175758
14        Wilmington    DE                  154            3.363636
15  Saint Petersburg    FL                  139            3.571942
16          Franklin    TN                  136            3.323529
17          Meridian    ID                  125            3.352000
18    St. Petersburg    FL                  113            3.738938
19            Sparks    NV                  112            3.334821

First, we identified which states have the highest open_business_count and compared it with their average_stats_open for analysis. We can see that the top three areas with the most operating restaurants are Philadelphia, PA; Tampa, FL; and Indianapolis, IN. These three areas have the largest number of restaurants, while LA, which receives the most reviews, only ranks 13th. From this, we can infer that while PA and FL rank second and third in terms of review count, there is still room for growth in these regions, making them viable targets for promotion by Yelp and local restaurants. This finding also highlights the enthusiasm of LA residents for leaving reviews and underscores the importance of reviews for restaurant operations in that area.
Additionally, when comparing ratings, we observe that restaurants in California (CA) receive the highest ratings, reaching an average of 3.94, indicating a high level of satisfaction among local residents. LA also ranks high with an average of 3.7. On the other hand, areas like IN and AZ have relatively lower ratings, suggesting that restaurants in more urban areas may experience higher satisfaction levels.
Finally, when looking at closed restaurants, the distribution is largely similar to that of operating ones, indicating that market conditions are generally consistent.


<font face="Helvetica" size=5 color=2E0202> **Q10. What are the top attributes across businesses? How do they  impact ratings?** <font>

In [39]:
%%bigquery
SELECT
    attribute,
    COUNTIF(value = 'True') AS occurance_counts, 
    AVG(CASE WHEN value = 'True' THEN stars END) AS avg_stars,
    AVG(CASE WHEN value = 'True' THEN review_count END) AS avg_review_count
FROM (
    SELECT
        JSON_EXTRACT_SCALAR(attributes, '$.ByAppointmentOnly') AS ByAppointmentOnly,
        JSON_EXTRACT_SCALAR(attributes, '$.Caters') AS Caters,
        JSON_EXTRACT_SCALAR(attributes, '$.RestaurantsTableService') AS RestaurantsTableService,
        JSON_EXTRACT_SCALAR(attributes, '$.CoatCheck') AS CoatCheck,
        JSON_EXTRACT_SCALAR(attributes, '$.HairSpecializesIn') AS HairSpecializesIn,
        JSON_EXTRACT_SCALAR(attributes, '$.GoodForMeal') AS GoodForMeal,
        JSON_EXTRACT_SCALAR(attributes, '$.GoodForKids') AS GoodForKids,
        JSON_EXTRACT_SCALAR(attributes, '$.BestNights') AS BestNights,
        JSON_EXTRACT_SCALAR(attributes, '$.BusinessParking') AS BusinessParking,
        JSON_EXTRACT_SCALAR(attributes, '$.BYOBCorkage') AS BYOBCorkage,
        JSON_EXTRACT_SCALAR(attributes, '$.Open24Hours') AS Open24Hours,
        JSON_EXTRACT_SCALAR(attributes, '$.RestaurantsDelivery') AS RestaurantsDelivery,
        JSON_EXTRACT_SCALAR(attributes, '$.DriveThru') AS DriveThru,
        JSON_EXTRACT_SCALAR(attributes, '$.OutdoorSeating') AS OutdoorSeating,
        JSON_EXTRACT_SCALAR(attributes, '$.Alcohol') AS Alcohol,
        JSON_EXTRACT_SCALAR(attributes, '$.RestaurantsCounterService') AS RestaurantsCounterService,
        JSON_EXTRACT_SCALAR(attributes, '$.RestaurantsReservations') AS RestaurantsReservations,
        JSON_EXTRACT_SCALAR(attributes, '$.WheelchairAccessible') AS WheelchairAccessible,
        JSON_EXTRACT_SCALAR(attributes, '$.RestaurantsGoodForGroups') AS RestaurantsGoodForGroups,
        JSON_EXTRACT_SCALAR(attributes, '$.AcceptsInsurance') AS AcceptsInsurance,
        JSON_EXTRACT_SCALAR(attributes, '$.HappyHour') AS HappyHour,
        JSON_EXTRACT_SCALAR(attributes, '$.RestaurantsTakeOut') AS RestaurantsTakeOut,
        JSON_EXTRACT_SCALAR(attributes, '$.RestaurantsPriceRange2') AS RestaurantsPriceRange2,
        JSON_EXTRACT_SCALAR(attributes, '$.BikeParking') AS BikeParking,
        JSON_EXTRACT_SCALAR(attributes, '$.BusinessAcceptsBitcoin') AS BusinessAcceptsBitcoin,
        JSON_EXTRACT_SCALAR(attributes, '$.WiFi') AS WiFi,
        JSON_EXTRACT_SCALAR(attributes, '$.BYOB') AS BYOB,
        JSON_EXTRACT_SCALAR(attributes, '$.AgesAllowed') AS AgesAllowed,
        JSON_EXTRACT_SCALAR(attributes, '$.Smoking') AS Smoking,
        JSON_EXTRACT_SCALAR(attributes, '$.Ambience') AS Ambience,
        JSON_EXTRACT_SCALAR(attributes, '$.DietaryRestrictions') AS DietaryRestrictions,
        JSON_EXTRACT_SCALAR(attributes, '$.Music') AS Music,
        JSON_EXTRACT_SCALAR(attributes, '$.GoodForDancing') AS GoodForDancing,
        JSON_EXTRACT_SCALAR(attributes, '$.Corkage') AS Corkage,
        JSON_EXTRACT_SCALAR(attributes, '$.RestaurantsAttire') AS RestaurantsAttire,
        JSON_EXTRACT_SCALAR(attributes, '$.NoiseLevel') AS NoiseLevel,
        JSON_EXTRACT_SCALAR(attributes, '$.DogsAllowed') AS DogsAllowed,
        JSON_EXTRACT_SCALAR(attributes, '$.BusinessAcceptsCreditCards') AS BusinessAcceptsCreditCards,
        JSON_EXTRACT_SCALAR(attributes, '$.HasTV') AS HasTV,
        stars,
        review_count
    FROM `ba775-project-a04.yelp_data.restaurant_businesses`
)
UNPIVOT(value FOR attribute IN (
    ByAppointmentOnly, Caters, RestaurantsTableService, CoatCheck, HairSpecializesIn, GoodForMeal,
    GoodForKids, BestNights, BusinessParking, BYOBCorkage, Open24Hours, RestaurantsDelivery,
    DriveThru, OutdoorSeating, Alcohol, RestaurantsCounterService, RestaurantsReservations,
    WheelchairAccessible, RestaurantsGoodForGroups, AcceptsInsurance, HappyHour, RestaurantsTakeOut,
    RestaurantsPriceRange2, BikeParking, BusinessAcceptsBitcoin, WiFi, BYOB, AgesAllowed,
    Smoking, Ambience, DietaryRestrictions, Music, GoodForDancing, Corkage, RestaurantsAttire,
    NoiseLevel, DogsAllowed, BusinessAcceptsCreditCards, HasTV
))
GROUP BY attribute
ORDER BY occurance_counts DESC
LIMIT 10;


Query is running:   0%|          |

Downloading:   0%|          |

attribute  occurance_counts  avg_stars  avg_review_count
0          RestaurantsTakeOut             18403   3.445335         80.646199
1  BusinessAcceptsCreditCards             17941   3.425645         86.015049
2                 GoodForKids             13746   3.349447         94.599156
3    RestaurantsGoodForGroups             12586   3.369101        103.150802
4         RestaurantsDelivery             12252   3.236247         82.042116
5                 BikeParking             11073   3.620880        113.328998
6                       HasTV             11062   3.312376         90.331043
7                      Caters              8527   3.671045        106.010789
8              OutdoorSeating              8117   3.646175        116.471356
9        WheelchairAccessible              5297   3.934208        138.751557

According to the above analysis, the most popular attributes that businesses may declare as important are RestaurantsTakeOut, BusinessAcceptsCreditCards, GoodForKids. These top attributes make sense for being poplular as most individuals may or not want to eat at the establishment and prefer to have the capacity to take out or deliver the food for convinience. The might also value having varied accessibility at the restaurant like Wheelchair Accibility, bike parking, kid-friendly etc.

As far as the ratings and average review count goes, we see that wheel chair accessible locations have the highest average ratings and average review count. Wheelchair-accessible locations often receive higher ratings and review counts because they create an inclusive experience that appeals to a broader customer base, including families, seniors, and individuals with mobility needs. This commitment to accessibility enhances brand perception, leading customers to view these businesses as considerate and community-minded, which can encourage positive feedback. Additionally, accessible locations may maintain higher operational standards overall, contributing to a better experience for all customers and driving more frequent, favorable reviews.

Out of the 10 most popular attributes, RestaurantsDelivery has the lowest score and lowest review count. One reason for the low scores and review counts for "RestaurantsDelivery" could be that delivery services often don't provide the full dining experience, which can lead to lower satisfaction. Delivery orders may lack the ambiance, presentation, and quality control of in-restaurant dining, which could result in a less favorable experience and lower ratings. Additionally, delivery experiences may be more likely to go unreviewed, as customers are less engaged with the business itself, contributing to the lower review count.


<font face="Helvetica" size=6 color=B40E0E> **Conclusion<a name="conclusions"></a>** <font>

----
    
    
The analysis highlights Yelp’s growth in restaurant reviews until a COVID-driven decline in 2020, with elite users continuing to influence the platform significantly through high engagement and quality contributions. Wheelchair-accessible locations have the highest ratings and review counts, reflecting their appeal to a broad audience. McDonald’s leads in customer loyalty with repeated reviews, though it holds lower average ratings, contrasting with smaller, high-quality local establishments that enjoy higher ratings. Review activity peaks on Sundays, Wednesdays, and Saturdays, underscoring key service days for restaurants. Patterns also reveal that low-star reviews are often both humorous and detailed, while 4- and 5-star reviews are generally less informative. Key restaurant attributes include takeout and kid-friendliness, while delivery receives lower ratings, likely due to a reduced dining experience. Lastly, restaurants with high review counts generally maintain positive ratings, while those with fewer reviews show lower satisfaction.

<font face="Helvetica" size=5 color=2E0202> **Recommendations <a name="Recommendations"></a>** <font>
    
---
    
- **Focus on Quality Over Quantity:** Restaurant owners should concentrate on delivering high-quality service at a limited number of locations to build a strong, loyal customer base. Satisfied customers will drive positive word-of-mouth and attract new patrons. Once a consistent high standard is achieved, expansion or franchising could be considered to broaden the customer base while maintaining quality.
    
- **Enhance User Engagement:** Yelp could encourage users to add more details to 4- and 5-star reviews, making positive feedback more informative and valuable to readers. A dedicated "Top Cool Reviews" section for popular 4-star reviews would make the platform feel more relatable, while a "Funny Reviews" section for amusing low-star reviews would add enjoyment to browsing.
    
- **Provide Business Insights:** Yelp could help businesses improve by showing which reviews are marked as "useful" or "cool." This feedback would allow businesses to better understand what resonates with customers and make targeted improvements.

- **Boost Elite User Contributions:** To keep elite users engaged, Yelp could provide regular feedback on how their reviews benefit the community. Businesses could publicly thank elite users on Yelp for helpful reviews, fostering loyalty. Public recognition for elite users who reach specific milestones (with badges or levels) would further motivate quality contributions. Additionally, promoting the perks of elite status among non-elite users could encourage broader participation.

<font face="Helvetica" size=5 color=2E0202> **Future Steps <a name="Future Steps"></a>** <font>
    
---
For our future steps, we would like to analyse the following:
    
- **Urban vs. Rural Satisfaction:** Assess how location type (urban or rural) impacts customer satisfaction and examine reviews for unique insights related to each area. This could reveal differences in expectations and preferences, guiding businesses in tailoring their approach based on location.

- **Review Length and Rating Correlation:** Explore the relationship between review length and rating type (positive or negative) to understand how review depth correlates with customer sentiment.
    
- **COVID-19 Review Trends:** Explore general trends in customer reviews during the COVID-19 pandemic to understand shifts in customer expectations and experiences, which may still be influencing the industry today.

<font face="Helvetica" size=6 color=B40E0E> **Challenges<a name="challenges"></a>** <font>

----
    
We faced the following challenges with the project:

- **Extracting Weekly Trends from Date Column**: The date column is formatted by calendar date (MM-DD-YYYY), making it challenging to extract specific days of the week. To address this, the `FORMAT_DATE` function was used to determine the number of reviews per day of the week across all dataset years.

- **Separating Years in Elite Column**: The elite column contains multiple years as numbers, requiring conversion to string format and use of `REGEXP_EXTRACT` to separate each year by commas.

- **Complex Category Structure**: Categories include multiple subcategories separated by commas, complicating the identification of a single representative category. This part was excluded from the current analysis and will be addressed in further analysis.

- **Diverse Attribute Labels**: Numerous varied labels within attributes add complexity to the analysis.

- **Data Table Creation Errors**: Errors were encountered during data table creation, particularly with the business table. These were resolved by using advanced options, such as increasing the allowed error count and ignoring unknown values.

<font face="Helvetica" size=6 color=B40E0E> **References<a name="reference"></a>** <font>

----
    
- https://www.eater.com/2019/1/16/18184376/restaurants-disabled-accessibility-americans-with-disabilities-act#:~:text=Disabled%20people%20have%20the%20money,experience%20from%20start%20to%20finish.
- https://www.washingtonpost.com/lifestyle/food/why-i-will-start-including-accessibility-information-in-my-restaurant-reviews/2019/05/10/b25560aa-7014-11e9-9eb4-0828f5389013_story.html
- https://www.zionandzion.com/research/food-delivery-apps-usage-and-demographics-winners-losers-and-laggards/#:~:text=Food%20delivery%20websites/apps%20suffer,delivery%20website/app%20feedback%20process.

<font face="Helvetica" size=6 color=B40E0E> **Generative AI Discolure<a name="genai"></a>** <font>

----

We have utilized Generative AI tools to assist with various aspects of this project. Below describe the criteria we used the tools:


* **Code Debugging:** Intellicode and ChatGPT suggested ways to improve our SQL queries, including providing example code, optimizing algorithms, and resolving potential errors.

* **Proofreading and Grammar Checks:** We used Grammarly and ChatGPT to refine our writing, improve readability, and ensure grammatical accuracy.

* **Fact Check for key observations:** Utilized ChatGPT as a resource to fact check assumptions and statements made in the analysis process.

* **Assistance with code:** Generative AI was utilized to correctly query attributes and hours columns from the business table, enabling the generation of keywords like UNPIVOT().


Our team has reviewed, edited, and validated all AI-generated content to ensure its accuracy, relevance, and originality in accordance with academic integrity guidelines.